images/cfw/opt/x1plus/lib/python/x1plus/services/x1plusd/modules/led_animations/states.py

```diff
@@ -0,0 +1,152 @@
+                self.leds.put([int(a * 255) for a in arr])
+                await asyncio.sleep(on / 1000.0)
+                
+                self.leds.put(b'\x00\x00\x00' * self.leds.n_leds)
+                await asyncio.sleep(off / 1000.0)
+ 
+
+class FinishAnimation():
+    def __init__(self, leds, config):
+        self.leds = leds
+        self.brightness = config.get('brightness', 0.4)
+        self.timeout = config.get('timeout', 600) # success goes away after 10 minutes
+        self.last_finish_trn = 0
+        self.last_was_finish = False
+    
+    def can_render(self):
+        finish = self.leds.last_gcode_state == 'FINISH'
+        if not self.last_was_finish and finish:
+            self.last_finish_trn = time.time()
+        self.last_was_finish = finish
+        return finish and (time.time() - self.last_finish_trn) < self.timeout
+    
+    async def task(self):
+        ph = 0
+
+        while True:
+            ph += 0.05
+            br = 0.6 + math.sin(ph) * 0.4
+            self.leds.put([int(255 * br * self.brightness), 0, 0] * self.leds.n_leds)
+            await asyncio.sleep(0.05)
+
+
+class FailedAnimation():
+    def __init__(self, leds, config):
+        self.leds = leds
+        self.brightness = config.get('brightness', 0.25)
+        self.timeout = config.get('timeout', 120) # failure goes away after 2 minutes
+        self.last_failed_trn = 0
+        self.last_was_failed = False
+    
+    def can_render(self):
+        failed = self.leds.last_gcode_state == 'FAILED'
+        if not self.last_was_failed and failed:
+            self.last_failed_trn = time.time()
+        self.last_was_failed = failed
+        return failed and (time.time() - self.last_failed_trn) < self.timeout
+    
+    async def task(self):
+        BLINK_PATTERN = [(750, 350), (750, 350), (750, 3000)]
+        while True:
+            for on,off in BLINK_PATTERN:
+                arr = []
+                for i in range(self.leds.n_leds):
+                    arr += (0.0, self.brightness * 1.0, 0.0,)
+                self.leds.put([int(a * 255) for a in arr])
+                await asyncio.sleep(on / 1000.0)
+                
+                arr = []
+                for i in range(self.leds.n_leds):
+                    arr += (0.0, 1/255, 0.0,)
+                self.leds.put([int(a * 255) for a in arr])
+                await asyncio.sleep(off / 1000.0)
+
+
+@register_animation("running")
+def _animation_running(handler):
+    handler.ANIMATIONS.setdefault("running", RunningAnimation)
+   
+@register_animation("paused")
+def _animation_paused(handler):
+    handler.ANIMATIONS.setdefault("paused", PausedAnimation)
+    
+@register_animation("finish")
+def _animation_finish(handler):
+    handler.ANIMATIONS.setdefault("finish", FinishAnimation)
+    
+@register_animation("failed")
+def _animation_failed(handler):
+    handler.ANIMATIONS.setdefault("failed", FailedAnimation)
```

Comment:
hmm, why not just have these be class decorators?

Reply:
Was copy-pasting it since these were at some point separated out into individual files for consistency. They could be, but this way it's slightly more consistent with non-core animations that are loaded, plus was the decorator usage already in place before PR.